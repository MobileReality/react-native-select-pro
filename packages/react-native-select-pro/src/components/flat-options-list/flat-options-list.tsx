import React, { memo, useRef } from 'react';
import isEqual from 'react-fast-compare';
import { FlatList } from 'react-native';

import { ERRORS, logError } from '../../helpers';
import { NoOptions } from '../no-options';

import type { FlatOptionsListProps } from './flat-options-list.types';

export const FlatOptionsList = memo(
    <T,>({
        renderItem,
        getItemLayout,
        resolvedData,
        flatListProps,
        initialScrollIndex,
        accessibilityState,
        disabled,
    }: FlatOptionsListProps<T>) => {
        const flatListRef = useRef<FlatList>(null);

        const scrollToIndex = () => {
            if (flatListRef.current) {
                try {
                    flatListRef.current.scrollToIndex({
                        animated: false,
                        index: initialScrollIndex === -1 ? 0 : initialScrollIndex,
                    });
                } catch {
                    logError(ERRORS.SCROLL_TO_LOCATION);
                }
            }
        };

        return (
            <FlatList
                testID="Options list"
                accessibilityLabel="Options list"
                accessibilityState={accessibilityState}
                bounces={false}
                keyboardShouldPersistTaps="handled"
                persistentScrollbar={true}
                ListEmptyComponent={<NoOptions />}
                scrollEnabled={!disabled}
                {...flatListProps}
                data={resolvedData}
                getItemLayout={getItemLayout}
                renderItem={renderItem}
                keyExtractor={({ value }) => value}
                onLayout={scrollToIndex}
            />
        );
    },
    isEqual,
);

FlatOptionsList.displayName = 'FlatOptionsList';
