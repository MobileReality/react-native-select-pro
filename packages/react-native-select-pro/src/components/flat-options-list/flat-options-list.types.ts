import type { FlatListProps } from 'react-native';

import type { OptionalToRequired, OptionType, SelectProps } from '../../types';

export type FlatOptionsListProps<T> = {
    resolvedData: OptionType<T>[];
    renderItem: FlatListProps<OptionType<T>>['renderItem'];
    getItemLayout: FlatListProps<OptionType<T>>['getItemLayout'];
    accessibilityState: FlatListProps<OptionType<T>>['accessibilityState'];
    initialScrollIndex: number;
} & OptionalToRequired<Pick<SelectProps<T>, 'flatListProps' | 'disabled'>>;
