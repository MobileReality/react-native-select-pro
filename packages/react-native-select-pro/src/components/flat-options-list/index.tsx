export { FlatOptionsList } from './flat-options-list';
