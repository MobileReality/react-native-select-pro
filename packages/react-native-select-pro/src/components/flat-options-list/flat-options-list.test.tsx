import React from 'react';
import { Text } from 'react-native';
import { render } from '@testing-library/react-native';

import { FlatOptionsList } from './flat-options-list';

describe('FlatOptionsList', () => {
    it('should generate FlatOptionsList snapshot', () => {
        const DATA = [{ label: 'test', value: 'test' }];
        const ITEM_LAYOUT = {
            length: 0,
            offset: 0,
            index: 0,
        };
        const flatOptionsList = render(
            <FlatOptionsList
                flatListProps={undefined}
                isOpened={false}
                noOptionsText={undefined}
                scrollToSelectedOption={false}
                NoOptionsComponent={undefined}
                selectedOptionIndex={0}
                resolveData={() => DATA}
                renderItem={() => <Text>Test Item</Text>}
                getItemLayout={() => ITEM_LAYOUT}
                onPressOption={() => {}}
            />,
        );
        expect(flatOptionsList).toMatchSnapshot();
    });
});
