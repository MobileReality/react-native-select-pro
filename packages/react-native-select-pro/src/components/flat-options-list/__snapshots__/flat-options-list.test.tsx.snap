// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlatOptionsList should generate FlatOptionsList snapshot 1`] = `
<RCTScrollView
  ListEmptyComponent={<NoOptions />}
  accessibilityLabel="Options list"
  accessibilityState={
    {
      "expanded": false,
    }
  }
  bounces={false}
  data={
    [
      {
        "label": "test",
        "value": "test",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  getItemLayout={[Function]}
  keyExtractor={[Function]}
  keyboardShouldPersistTaps="handled"
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  persistentScrollbar={true}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={[]}
  testID="Options list"
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      style={null}
    >
      <Text>
        Test Item
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
